File: LeetCode/Medium/0046_Permutations/Solution.java
// 2022-06-22 (recursive backtracking)
class Solution {
    
    private List<List<Integer>> output;
    
    private void permDigit(List<Integer> nums, int spot, List<Integer> branch) {
        
        // base-case; if we've seen every number, then add the branch to our output.
        if (nums.size() == 0) 
            output.add(branch);
        
        // create a new branch if we're currently on the first digit place
        if (spot == 0) branch = new ArrayList<Integer>();
        
        // iterate every number that's available; send clones with the taken number into recursion 
        for (int i = 0; i < nums.size(); i++) {
                
                // cloning each element before sending down
                List<Integer> copybranch = new ArrayList<Integer>(branch);
                List<Integer> copylist = new ArrayList<Integer>(nums);
                
                // adding/removing used number to clones
                copybranch.add(nums.get(i));
                copylist.remove(nums.get(i));
                
                permDigit(copylist, spot + 1, copybranch);
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        
        // convert nums into convertable form for arraylist
        Integer[] output1 = Arrays.stream(nums).boxed().toArray(Integer[]::new);
        
        // initialize assets
        output = new ArrayList<List<Integer>>();
        List<Integer> numlist = Arrays.asList(output1);
        
        // recursion
        permDigit(numlist, 0, null);
        
        return output;
    }
}

File: LeetCode/Medium/0046_Permutations/Solution.py
from typing import *

## 2022-11-09 (recursive backtracking top-down with extra params)
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        output = []
        def findPerms(bag, current, output):
            if len(current) == len(bag):
                output.append(current[:])
                return
            
            for i in range(len(bag)):
                if bag[i] == 11:
                    continue
                current.append(bag[i])
                hold, bag[i] = bag[i], 11
                findPerms(bag, current, output)
                current.pop()
                bag[i] = hold
        
        findPerms(nums, [], output)
        return output

## 2022-11-09 (recursive backtracking bottom-up)
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        output = []
        for i in range(len(nums)):
            if nums[i] == 11:
                continue
            
            hold, nums[i] = nums[i], 11
            vals = self.permute(nums)
            for val in vals:
                val.append(hold)
                output.append(val)
            nums[i] = hold

        return output if output else [[]]